Adds support for advanced features to be chosen inside flags

Marcellofabrizio opened the pull request:
By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

## Problem/Feature

Implementation of #183

## Description of Changes: 

- I added flags for the advanced features to be added without opening the prompt
- I replaced "AddHTMXTmpl" with "Htmx" to be more standard with the other flags. For this, I replaced all files that used the old name.
- I replaced the hardcoded calls for "AddHTMXTmpl" and "GitHubAction" with their respective new flag constants

## Checklist

- [x] I have self-reviewed the changes being requested
- [x] I have updated the documentation (if applicable)

PS: I'm new to Go, so I would appreciate constructive feedback 😄

briancbarrow commented on cmd/create.go:
```diff
@@ -170,18 +172,35 @@ var createCmd = &cobra.Command{
-			if err != nil {
-				log.Fatal("failed to set the htmx option", err)
+
+			featureFlags := cmd.Flag("feature").Value.String()
+			hasFeatureFlags := featureFlags != ""
+
+			if hasFeatureFlags {
```
```suggestion
			featureFlags := cmd.Flag("feature").Value.String()

			if featureFlags != "" {
```

We can remove the extra variable and just do the check right on `featureFlags`

Marcellofabrizio replied:
Commited!

briancbarrow commented on cmd/create.go:
```diff
@@ -170,18 +172,35 @@ var createCmd = &cobra.Command{
+					project.AdvancedOptions[key] = true
+				}
+
+				fmt.Println(project.AdvancedOptions)
```
forgotten print statement 😃

Marcellofabrizio replied:
The best debbuger

briancbarrow commented on cmd/create.go:
```diff
@@ -170,18 +172,35 @@ var createCmd = &cobra.Command{
+				}
+				project.ExitCLI(tprogram)
+				for key, opt := range options.Advanced.Choices {
+					fmt.Println(key)
+					fmt.Println(opt)
```
forgotten print statements

briancbarrow commented on cmd/create.go:
```diff
@@ -228,7 +247,7 @@ var createCmd = &cobra.Command{
 		fmt.Println(endingMsgStyle.Render("\nNext steps:"))
 		fmt.Println(endingMsgStyle.Render(fmt.Sprintf("• cd into the newly created project with: `cd %s`\n", project.ProjectName)))
 
-		if options.Advanced.Choices["AddHTMXTempl"] {
+		if options.Advanced.Choices[string(flags.Htmx)] {
```
I like this change. 👍

briancbarrow commented on cmd/flags/advancedFeatures.go:
```diff
@@ -0,0 +1,36 @@
+		}
+	}
+
+	return fmt.Errorf("Advanced Feature to use. Allowed values: %s", strings.Join(AllowedAdvancedFeatures, ", "))
```
Go will yell at you if you have error strings that are capitalized. 🤷‍♂️ 
```suggestion
	return fmt.Errorf("advanced feature allowed values: %s", strings.Join(AllowedAdvancedFeatures, ", "))
```

Marcellofabrizio replied:
Commited!

Marcellofabrizio commented:
Hey @briancbarrow, I really appreciate the feedback, I'll be working on those comments as soon as I can. Regarding the suggested command, what should be the expected output?

briancbarrow commented:
> Hey @briancbarrow, I really appreciate the feedback, I'll be working on those comments as soon as I can. Regarding the suggested command, what should be the expected output?

We should either change the code to accept it like 
```
go-blueprint create --advanced true --feature htmx,githubaction
```

or like this (I think this will be easier to do).
```
go-blueprint create --advanced --feature htmx --feature githubaction
```

Marcellofabrizio commented:
@briancbarrow I implemented some changes that will allow us to print the non-interactive command correctly to the new format. Here's how it's looking:
![image](https://github.com/Melkeydev/go-blueprint/assets/39314230/95dc3cef-1b52-4950-9b5c-8f91c2201b92)
I took the liberty to remove the `true` following the `--advanced` as it looked redundant since we can't have the flag with a false value.

I also addressed your other comments.

Marcellofabrizio commented:
Hey @briancbarrow, I implemented the changes to print the correct command based on the new flags. Here's how it looks: 
![image](https://github.com/Melkeydev/go-blueprint/assets/39314230/5ad45fd3-123f-4872-97f0-4839c9cc9c65)

I also addressed your other comments.

Marcellofabrizio commented on cmd/create.go:
```diff
@@ -170,18 +172,35 @@ var createCmd = &cobra.Command{
+
+			if hasFeatureFlags {
+				featuresFlagValues := strings.Split(featureFlags, ",")
+				fmt.Println(len(featuresFlagValues))
```
The best debbuger

briancbarrow commented on cmd/create.go:
```diff
@@ -228,7 +240,7 @@ var createCmd = &cobra.Command{
 		fmt.Println(endingMsgStyle.Render(fmt.Sprintf("• cd into the newly created project with: `cd %s`\n", project.ProjectName)))
 
-		if options.Advanced.Choices["AddHTMXTempl"] {
+		if options.Advanced.Choices[string(flags.Htmx)] {
```
I was running through this again. I like the idea of this change, but the choice is `"Htmx"` and `string(flags.Htmx)` returns `"htmx"` so these 'next steps' never get printed. 
```suggestion
		if options.Advanced.Choices["Htmx"] {
```

briancbarrow commented:
> Hey @briancbarrow, I implemented the changes to print the correct command based on the new flags. Here's how it looks: ![image](https://private-user-images.githubusercontent.com/39314230/308382291-5ad45fd3-123f-4872-97f0-4839c9cc9c65.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk2OTg1MjAsIm5iZiI6MTcwOTY5ODIyMCwicGF0aCI6Ii8zOTMxNDIzMC8zMDgzODIyOTEtNWFkNDVmZDMtMTIzZi00ODcyLTk3ZjAtNDgzOWM5Y2M5YzY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA2VDA0MTAyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1NWViNmNhZWVmZjFiNmEwYzdmYzJmNjUwNzkwOTdjYzc4NzkyMjZkZTI1MTdlMjQ2NmI0NTk2OTI2ZmEwZTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mJCftZGcPS6fkkhMpnMA2HaVkVzKTSZp-95-TRD9Rlg)
> 
> I also addressed your other comments.

Sorry again for the late reply. Things are looking good. I did notice that now the steps for getting the htmx stuff working/generated are not printing as expected. See my comment above for that. I'm also still not getting the right output. I'm still getting 
`go-blueprint create --name my-proj --framework chi --driver mysql --advanced true` instead of the advanced features broken out. Can you confirm on your side that you are getting what is expected

<img width="698" alt="Screen Shot 2024-03-05 at 9 11 05 PM" src="https://github.com/Melkeydev/go-blueprint/assets/13695093/dbbbb3db-f6b2-4796-819a-5f67e959ae56">

Marcellofabrizio commented:
- @briancbarrow Yes, I can replicate this. You're getting `go-blueprint create --name my-proj --framework chi --driver mysql --advanced true` because my code is only handling the features added in flags, not the ones selected by the steps. I didn't consider this case when I added it. I'll change it to handle both cases.

Marcellofabrizio commented:
Hey @briancbarrow, it was a pretty simple fix, all I had to do was apply the same logic we are doing to the other flags, which is setting the selected value to the flag after we select them.

```
for key, opt := range options.Advanced.Choices {
	project.AdvancedOptions[strings.ToLower(key)] = opt
	err := cmd.Flag("feature").Value.Set(strings.ToLower(key))
	if err != nil {
		log.Fatal("failed to set the feature flag value", err)
	}
}
```
This allows the values to be passed down to `NonInteractiveCommand` and be printed as expected.

briancbarrow reviewed (approved):
This looks good to me now.

Marcellofabrizio commented:
@briancbarrow Thank you!

briancbarrow commented:
> @briancbarrow Thank you!

Thank you for working on this!
